python/ray/spark/__init__.py

```diff
@@ -0,0 +1,530 @@
+            if num_spark_tasks_for_heap_mem_req > num_spark_tasks:
+                num_spark_tasks = num_spark_tasks_for_heap_mem_req
+
+        if total_object_store_memory_bytes is not None:
+            if total_object_store_memory_bytes <= 0:
+                raise ValueError(
+                    "You should specify 'total_object_store_memory_bytes' argument to a positive integer."
+                )
+            num_spark_tasks_for_object_store_mem_req = \
+                int(math.ceil(total_object_store_memory_bytes / ray_worker_object_store_mem_bytes))
+            if num_spark_tasks_for_object_store_mem_req > num_spark_tasks:
+                num_spark_tasks = num_spark_tasks_for_object_store_mem_req
+
+    if num_spark_task_cpus < 4:
+        _logger.warning(
+            f"Each ray worker node will be assigned with {num_spark_task_cpus} CPU cores, less than "
+            "recommended value 4, because ray worker node cpu cors aligns with the cpu cores assigned to "
+            "a spark task, you can increase 'spark.task.cpus' config value to address it."
+        )
+
+    if ray_worker_heap_mem_bytes < 10 * 1024 * 1024 * 1024:
+        _logger.warning(
+            f"Each ray worker node will be assigned with {ray_worker_heap_mem_bytes} bytes heap memory, "
+            "less than recommended value 10GB, the ray worker node heap memory size is calculated by "
+            "(SPARK_WORKER_NODE_PHYSICAL_MEMORY - SHARED_MEMORY) / num_local_spark_task_slots * 0.8, "
+            "so you can increase spark cluster worker machine memory, or reduce spark task slots "
+            "number on spark cluster worker, or reduce spark worker machine /dev/shm quota to "
+            "address it."
+        )
+
+    ray_head_hostname = get_spark_driver_hostname(spark.conf.get("spark.master"))
+    ray_head_port = get_safe_port()
+
+    ray_head_node_manager_port = get_safe_port()
+    ray_head_object_manager_port = get_safe_port()
+
+    _logger.info(f"Ray head hostname {ray_head_hostname}, port {ray_head_port}")
+
+    ray_exec_path = os.path.join(os.path.dirname(sys.executable), "ray")
+
+    ray_log_dir = os.path.join(ray_log_dir, f"ray-{ray_head_port}")
+    os.makedirs(ray_log_dir, exist_ok=True)
+
+    # TODO: Many ray processes logs are outputted under "{ray_temp_dir}/session_latest/logs",
+    #  Proposal: Update "ray start" scirpt to add a new option "ray_log_dir", and output logs
+    #  to a different directory specified by "ray_log_dir", instead of using
+    #  "{ray_temp_dir}/session_latest/logs".
+    #  The reason is, for ray on spark, user is hard to access log files on spark worker machines,
+    #  (especially on databricks runtime), so we'd better set the log output dir to be a
+    #  path mounted with NFS shared by all spark cluster nodes, so that the user can access
+    #  these remote log files from spark drive side easily.
```

Comment:
@jjyao 

Can we support this for `ray start` CLI ?

>     #  Proposal: Update "ray start" scirpt to add a new option "ray_log_dir", and output logs
>     #  to a different directory specified by "ray_log_dir", instead of using
>     #  "{ray_temp_dir}/session_latest/logs".

Reply:
@jjyao Unfortunately the ticket does not address my demands: https://github.com/ray-project/ray/issues/28876#issuecomment-1322016494